### PR TITLE
feat<Compiler>: consider that the dispatch function from `useReducer` is non-reactive

### PR DESCRIPTION
**Pull request by TrickyPi**
## Summary

The dispatch function from `useReducer` is stable, so it is also non-reactive.

the related PR: https://github.com/facebook/react/pull/29665
the related comment: https://github.com/facebook/react/issues/29674#issuecomment-2141432192

I am not sure if the location of the new test file is appropriate😅.

## How did you test this change?
Added the specific test `compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/useReducer-returned-dispatcher-is-non-reactive.expect.md`.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts [hunk 1]
```diff
@@ -265,6 +266,18 @@ const REACT_APIS: Array<[string, BuiltInType]> = [
...
+  [
+    "useReducer",
+    addHook(DEFAULT_SHAPES, {
+      positionalParams: [],
+      restParam: Effect.Freeze,
+      returnType: { kind: "Object", shapeId: BuiltInUseReducerId },
+      calleeEffect: Effect.Read,
+      hookKind: "useReducer",
+      returnValueKind: ValueKind.Frozen,
+      returnValueReason: ValueReason.State,
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts:278]: Let’s add a new ValueReason.Reducer enum variant and use it here, so that we can customize the error message for reducers

File: compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts [hunk 2]
```diff
@@ -2116,7 +2116,7 @@ function getWriteErrorReason(abstractValue: AbstractValue): string {
   } else if (abstractValue.reason.has(ValueReason.ReactiveFunctionArgument)) {
     return "Mutating component props or hook arguments is not allowed. Consider using a local variable instead";
   } else if (abstractValue.reason.has(ValueReason.State)) {
-    return "Mutating a value returned from 'useState()', which should not be mutated. Use the setter function to update instead";
+    return "Mutating a value returned from 'useState() or useReducer()', which should not be mutated. Use the setter or dispatcher function to update instead";
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Inference/InferReferenceEffects.ts:2119]: See above, let’s create a new Reducer variant and customize the error message for it.

**Review by josephsavona — COMMENTED** [review]
Exciting, thanks for putting up a PR! Just some quick feedback, will do a full review next week

File: compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts (hunk 1 above)
Comment (TrickyPi) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/Globals.ts]: Done!

**Comment by TrickyPi** [comment]
Thanks for your review! The required changes have been made.

File: compiler/apps/playground/components/TabbedWindow.tsx
Comment (josephsavona) [inline compiler/apps/playground/components/TabbedWindow.tsx:1]: can you revert the changes here?

**Review by josephsavona — APPROVED** [review]
awesome, thanks!

**Review by josephsavona — CHANGES_REQUESTED** [review]
Looks like you need to run prettier

**Comment by TrickyPi** [comment]
Oh, I forgot. Now it's fixed.

**Comment by josephsavona** [comment]
Awesome, thanks!
